src/atom_space_node/AtomSpaceNode.h

```diff
@@ -0,0 +1,193 @@
+
+
+    /**
+     * Returns this node's ID.
+     *
+     * @return this node's ID.
+     */
+    string node_id();
+
+    /**
+     * Broadcasts a command to all nodes in the network.
+     *
+     * All nodes in the network will be reached (not only the known peers) and the command
+     * will be executed.
+     *
+     * @param command The command to be executed in the target nodes.
+     * @param args Arguments for the command.
+     */
+    void broadcast(const string &command, const vector<string> &args);
+
+    /**
+     * Sends a command to the passed node.
+     *
+     * The target node is supposed to be a known peer. If not, an exception is thrown.
+     *
+     * @param command The command to be executed in the target nodes.
+     * @param args Arguments for the command.
+     * @recipient The target node for the command.
+     */
+    void send(const string &command, const vector<string> &args, const string &recipient);
+
+    // --------------------------------------------------------------------------------------------
+    // API to be extended by concrete subclasses
+
+    /**
+     * This method is called whenever a new node joins the network.
+     *
+     * Every node that joins the network will broadcast a Message to everyone already joined.
+     * However, it's up to each node to decide whether or not the newly joined node will be added
+     * as a known peer.
+     *
+     * @param node_id The ID of the newly joined node.
+     */
+    virtual void node_joined_network(const string &node_id) = 0;
+
+    /**
+     * Casts a vote (which is the ID of the node being voted) in a leadership election.
+     *
+     * @return This node's vote (which is the ID of the node being voted) for leader.
+     */
+    virtual string cast_leadership_vote() = 0;
+
+protected:
+
+    // --------------------------------------------------------------------------------------------
+    // Protected API
+
+    /**
+     * Basic constructor.
+     *
+     * @param node_id The ID of this node. Typically is a string to identify this node
+     * in the MessageBroker.
+     * @param leadership_algorithm The concrete class to be used as leadership broker.
+     * @param messaging_backend The concrete class to be used as message broker.
+     */
+    AtomSpaceNode(
+        const string &node_id,
+        LeadershipBrokerType leadership_algorithm,
+        MessageBrokerType messaging_backend
+    );
+
+    /**
+     * Build the Message object which is supposed to execute the passed command.
+     *
+     * @param command The command to be remotely executed.
+     * @param args Arguments for the command.
+     * @return A Message object
+     */
+    virtual Message *message_factory(string &command, vector<string> &args);
+
+private:
+
+    struct {
+        string NODE_JOINED_NETWORK = "node_joined_network";
+    } known_commands;
+
+    string my_node_id;
+    LeadershipBroker *leadership_broker;
+    MessageBroker *message_broker;
```

Comment:
I know we should have in mind performance and should avoid unnecessary overheads, but I'd like to see these pointers being handled by _smart pointers_ (`unique_ptr`, `shared_ptr`, ...) instead of being declared as raw pointers (requiring to be explicitly destroyed with `delete` statements). Not a big issue in this case, even because those attributes are private and constructed and destroyed just once per `AtomSpaceNode` instantiation - so, easy to manage.

Reply:
OK. Added a card to fix this: https://github.com/singnet/das-attention-broker/issues/28